### ISSUE TITLE
`dmypy` is not killed when python-lsp exits

### ISSUE THREAD

**Issue by itaranto** [python-lsp/pylsp-mypy#88]
I'm using Neovim with its native LSP client.

If I close the editor I still see an instance of `dmypy` up and running.

When editing a `.py` file I can see the follwing using `pstree -lp`:

![image](https://github.com/user-attachments/assets/8f96e543-bc37-41af-b21a-bfc5d8a43c22)

It seems to be there's one instance of `dmypy` which in fact is a child process of `pylsp`, this instance gets killed when I close my editor (and by extension, `pylsp` too).  But there's another instance which seems to be running in the background.

These are my settings:

```lua

    lspconfig.pylsp.setup {
      capabilities = capabilities,
      settings = {
        pylsp = {
          plugins = {
            autopep8 = { enabled = false },
            flake8 = { enabled = false },
            mccabe = { enabled = false },
            pycodestyle = { enabled = false },
            pydocstyle = { enabled = false },
            pyflakes = { enabled = false },
            pylint = { enabled = false },
            pylsp_mypy = {
              dmypy = true,
              enabled = true,
              live_mode = false,
            },
            yapf = { enabled = false },
          }
        }
      }
    }
```

**Comment by itaranto** [comment]
> Where do you get the `dmypy daemon` from? I can't find any reference to that in the documentation. 

```
[I] ~ $ dmypy --help
usage: dmypy [-h] [--status-file STATUS_FILE] [-V] {start,restart,status,stop,kill,check,run,recheck,suggest,inspect,hang,daemon,help} ...

Client for mypy daemon mode

positional arguments:
  {start,restart,status,stop,kill,check,run,recheck,suggest,inspect,hang,daemon,help}
    start               Start daemon
    restart             Restart daemon (stop or kill followed by start)
    status              Show daemon status
    stop                Stop daemon (asks it politely to go away)
    kill                Kill daemon (kills the process)
    check               Check some files (requires daemon)
    run                 Check some files, [re]starting daemon if necessary
    recheck             Re-check the previous list of files, with optional modifications (requires daemon)
    suggest             Suggest a signature or show call sites for a specific function
    inspect             Locate and statically inspect expression(s)
    hang                Hang for 100 seconds
    daemon              Run daemon in foreground

options:
  -h, --help            show this help message and exit
  --status-file STATUS_FILE
                        status file to retrieve daemon details
  -V, --version         Show program's version number and exit
```

> The deamon staying open is expected (this is what it is supposed to do). One could argue, that we should stop it atexit.

That could be a possibility, yes. It would be also better to run `dmypy` in such a way that's a child process of `pylsp` so when `pylsp` dies, `dmypy` dies too.

**Comment by Richardk2n** [comment]
Interestingly this is not in the [doc](https://mypy.readthedocs.io/en/stable/mypy_daemon.html). Also, `Run daemon in foreground` does not sound like the thing we want to me. Furthermore, we do run it either as a child process or in the same process (depending on venvs). It starting additional separated processes is a sensible design decision of `mypy`, which we cannot influence. Unless you can provide a working way to implement your preferred solution, I will take a look at stopping it `atexit`. BTW, what OS are you using? According to `mypy` src it might matter for this.

**Comment by itaranto** [comment]
> Also, `Run daemon in foreground` does not sound like the thing we want to me. Furthermore, we do run it either as a child process or in the same process (depending on venvs). 

Yes, it's confusing, I know. I just saw that option and ran it and then I saw how `dmypy` behaved. The results where that `dmypy` didn't spawn any "daemonized" processes, just the one started from the terminal.

>It starting additional separated processes is a sensible design decision of `mypy`, which we cannot influence. Unless you can provide a working way to implement your preferred solution, I will take a look at stopping it `atexit`. 

I can take a look there, yes. I can also file a bug/feature request to `mypy` to see if I get any response.

> BTW, what OS are you using? According to `mypy` src it might matter for this.

I'm running Linux.

**Comment by dlax** [comment]
In https://github.com/python-lsp/pylsp-mypy/pull/104, I suggest to leverage the `--timeout` option of `dmypy run` to make the daemon automatically shutdown after a specified inactivity timeout.

**Comment by Richardk2n** [comment]
Sorry I did not do anything for a year.
\<Insert currently doing my PhD excuse here>
Not sure if I want to restart the daemon every \<arbitrary amount of time>, that kinda defeats the purpose.
@dlax I guess you have the same issue?
Is there an argument against the `atexit` route that I am not seeing?

**Comment by dlax** [comment]
> Is there an argument against the `atexit` route that I am not seeing?

Can you clarify how you'd hook to `atexit`? Is there a pylsp hook for that?

**Comment by Richardk2n** [comment]
There is the python inbuilt `atexit` module, which should be called on a normal termination.
This would trigger, when the `pylsp` server stops.
For the editors I know, the server is started by the editor and therefore stops, if the editor is closed.
If someone manually starts a server, he is responsible for manually closing it as well.
This should mean, that having a `dmypy stop` in an `atexit` registered function aka here
https://github.com/python-lsp/pylsp-mypy/blob/cac3c40018777608bb9be3ad805742664a643b8d/pylsp_mypy/plugin.py#L657
 should do the trick.
(Assuming smooth shutdown)

**Comment by dlax** [comment]
Thanks for detailing your thought.

The solution probably works (I just hacked around [this](https://github.com/python-lsp/pylsp-mypy/commit/9ed4e25d0c220f83d785b15a2d107bc920813248)) but it's not very clean as far as I can tell because we don't have access to `settings` and thus would need to store at least `dmypy_command` and `dmypy_status_file` as globals. Probably better than nothing though...

On the other hand, I wonder if something shouldn't be done in pylsp so that we can get access to a hook upon [shutdown request](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#shutdown). This would be cleaner IMHO as we'd be able to avoid those globals.

**Comment by Richardk2n** [comment]
I know it is not that pretty and technically causes a bit of a regression (if there is cashing, the parameters cannot be changed during execution anymore). I am fine with this though. As I've been thinking about some abstraction, that would end in cashing anyway.

I guess there is an argument to be made for that and the implementation would not be too difficult.
Please ask in the server repo, if they would consider such a hook.

I will try to make time this weekend to try the abstraction I've been considering.

**Comment by Richardk2n** [comment]
Sorry, that I haven't worked on this yet.
Turns out, when I moved this repo to the org, I did not get any rights on it.
So I can't even create a branch, which I would very much like to do for a larger change.
I have not yet reached Carlos to fix this.
Thank you for opening the PR in the server repo.

**Comment by dlax** [comment]
The `pylsp_shutdown()` hook is now available in python-lsp-server (not yet released though); so I went on with this [commit](https://github.com/dlax/pylsp-mypy/commit/2460c40b0da60b2c251b9fa8c7debfe5fac18056) to implement dmypy stop (and tmpfile closing) this way. I'm using it regularly and it seems to work okay. Let me know if you want a PR for this.

**Comment by Richardk2n** [comment]
Looks good.
I might make some small adjustments.
Please make a PR, I will only merge it after the server has a new release (In case I have to do a hotfix).
Feel free to ping me once there is a release.

**Comment by dlax** [comment]
See #108. The required server version got released earlier: https://github.com/python-lsp/python-lsp-server/releases/tag/v1.13.1
